src/include/storage/compressed_tile.h

```diff
@@ -0,0 +1,165 @@
+
+  std::vector<type::Value> CompressColumn(
+      Tile *tile, oid_t column_id, std::vector<type::Value> column_values,
+      type::Value &base_value, type::Type::TypeId &compression_type);
+
+  std::vector<type::Value> CompressCharColumn(Tile *tile, oid_t column_id);
+
+  void InsertTuple(const oid_t tuple_offset, Tuple *tuple);
+
+  type::Value GetValue(const oid_t tuple_offset, const oid_t column_id);
+
+  type::Value GetValueFast(const oid_t tuple_offset, const size_t column_offset,
+                           const type::Type::TypeId column_type,
+                           const bool is_inlined);
+
+  void SetValue(const type::Value &value, const oid_t tuple_offset,
+                const oid_t column_id);
+
+  void SetValueFast(const type::Value &value, const oid_t tuple_offset,
+                    const size_t column_offset, const bool is_inlined,
+                    const size_t column_length);
+
+  type::Value GetUncompressedVarcharValue(oid_t column_id,
+                                          type::Value compressed_value);
+  //===--------------------------------------------------------------------===//
+  // Utility Functions
+  //===--------------------------------------------------------------------===//
+
+  inline bool IsCompressed() { return is_compressed; }
+
+  inline type::Type::TypeId GetCompressedType(type::Value new_value) {
+    return new_value.GetTypeId();
+  }
+
+  inline void SetCompressedMapValue(oid_t column_id, type::Type::TypeId type_id,
+                                    type::Value base_value) {
+    compressed_column_map[column_id] = std::make_pair(type_id, base_value);
+  }
+
+  inline void SetExponentMapValue(oid_t column_id, type::Value exponent) {
+    exponent_column_map[column_id] = exponent;
+  }
+
+  // this is for dictionary
+  inline void SetDecoderMapValue(oid_t column_id,
+                                 std::vector<type::Value> decoder) {
+    decoder_map[column_id] = decoder;
+  }
+
+  inline type::Value GetBaseValue(oid_t column_id) {
+    if (compressed_column_map.find(column_id) != compressed_column_map.end()) {
+      return compressed_column_map[column_id].second;
+    }
+    return type::Value();
+  }
+
+  inline type::Type::TypeId GetCompressedType(oid_t column_id) {
+    if (compressed_column_map.find(column_id) != compressed_column_map.end()) {
+      return compressed_column_map[column_id].first;
+    }
+    return type::Type::INVALID;
+  }
+
+  type::Value GetUncompressedValue(oid_t column_id,
+                                   type::Value compressed_value);
+
+  oid_t GetColumnFromOffset(const size_t column_offset) {
+    return column_offset_map[column_offset];
+  }
+
+ protected:
+  //===--------------------------------------------------------------------===//
+  // Data members
+  //===--------------------------------------------------------------------===//
+
+  bool is_compressed;
+  oid_t compressed_columns_count;
+  int tuple_count;
+
```

Comment:
I would suggest to add comments for these maps, such as what the key and value are.